libgadget/cooling_qso_lightup.c

```diff
@@ -0,0 +1,613 @@
+                                      the code will flash-ionize all remaining particles*/
+    double heIIIreion_start; /* Time at which start_reionization is called and helium III reionization begins*/
+};
+
+static struct qso_lightup_params QSOLightupParams;
+/* Memory for the helium reionization history. */
+/* Instantaneous heating from low-energy (short mean free path)
+ * photons to the Quasar to a newly ionized particle.
+ * Computed from parameters stored in the text file.
+ * In ergs.*/
+static double qso_inst_heating;
+static int Nreionhist;
+static double * He_zz;
+static double * XHeIII;
+static double * LMFP;
+static gsl_interp * HeIII_intp;
+static gsl_interp * LMFP_intp;
+/* Counter for particles ionized this step*/
+static int N_ionized;
+
+/*This is a helper for the tests*/
+void set_qso_lightup_par(struct qso_lightup_params qso)
+{
+    QSOLightupParams = qso;
+}
+
+void
+set_qso_lightup_params(ParameterSet * ps)
+{
+    int ThisTask;
+    MPI_Comm_rank(MPI_COMM_WORLD, &ThisTask);
+    if(ThisTask == 0) {
+        QSOLightupParams.QSOLightupOn = param_get_int(ps, "QSOLightupOn");
+        QSOLightupParams.qso_candidate_max_mass = param_get_double(ps, "QSOMaxMass");
+        QSOLightupParams.qso_candidate_min_mass = param_get_double(ps, "QSOMinMass");
+        QSOLightupParams.mean_bubble = param_get_double(ps, "QSOMeanBubble");
+        QSOLightupParams.var_bubble = param_get_double(ps, "QSOVarBubble");
+        QSOLightupParams.heIIIreion_start = param_get_double(ps, "QSOHeIIIReionStart");
+        QSOLightupParams.heIIIreion_finish_frac = param_get_double(ps, "QSOHeIIIReionFinishFrac");
+    }
+    MPI_Bcast(&QSOLightupParams, sizeof(struct qso_lightup_params), MPI_BYTE, 0, MPI_COMM_WORLD);
+}
+
+/* Instantaneous heat injection from HeII reionization
+ * (absorption of photons with E<Emax) per helium atom [ergs]
+ */
+static double
+Q_inst(double Emax, double alpha_q)
+{
+    /* Total ionizing flux for the short mean free path photons*/
+    double intflux = (pow(Emax,-alpha_q+1.)-pow(E0_HeII,-alpha_q+1.))/(pow(Emax,-alpha_q)- pow(E0_HeII,-alpha_q));
+    /* Heating is input per unit mass, so quasars in denser areas will provide more heating.*/
+    double Q_inst = (alpha_q/(alpha_q - 1.0))*intflux -E0_HeII;
+    return eVinergs * Q_inst;
+}
+
+/* Load in reionization history file and build the interpolators between redshift and XHeII.
+ * Format is:
+ * Need to load a text file with the shape of reionization, which gives the overall HeIII fraction,
+ * and the uniform background heating rate. Includes quasar spectral index, etc and most of the physics.
+ * Be careful. Do not double-count uniform long mean free path photons with the homogeneous UVB.
+ *
+ * quasar spectral index
+ * instantaneous absorption threshold energy (in eV)
+ * table of 3 columns, redshift, HeIII fraction, uniform background heating.
+ * The text file specifies the end redshift of reionization.
```

Comment:
Perhaps also list the first few lines of the file as an example, or make add a link to a real file. The comments alone still seem to be a bit cryptic.

Reply:
Also, is there a tool to generate the file? If so, perhaps also mention it here.